### PR TITLE
Add PHP 7 argument & return type declarations to all member methods

### LINKED ISSUE #363: Make use of PHP 7 type declarations
We require PHP 7 for Icinga DB Web, so let's make use of the new [type declarations](https://www.php.net/manual/de/migration70.new-features.php).

This is mainly about:
* Scalar type declarations
* Return type declarations

We should use them for our own methods everywhere.

### PR DESCRIPTION
**Pull request by Yonas-net**
resolves #363

### CODE REVIEW

File: library/Icingadb/Compat/UrlMigrator.php [hunk 1]
```diff
@@ -39,18 +39,18 @@ class UrlMigrator
...
 
-    public static function isSupportedUrl(Url $url)
+    public static function isSupportedUrl(Url $url): bool
     {
         $supportedPaths = self::SUPPORTED_PATHS;
         return isset($supportedPaths[$url->getPath()]);
     }
 
-    public static function hasQueryTransformer($name)
+    public static function hasQueryTransformer($name): bool
```
Comment (sukhwinder33445) [inline library/Icingadb/Compat/UrlMigrator.php:48]: Is $name not a string here?

File: application/forms/RedisConfigForm.php [hunk 2]
```diff
@@ -232,7 +232,7 @@ public static function addInsecureCheckboxIfTls(Form $form)
         $displayGroup->setElements($elements);
     }
 
-    public function isValid($formData)
+    public function isValid($formData): bool
```
[the remark this answers is not in the document]
Comment (Yonas-net) [reply application/forms/RedisConfigForm.php:235]: This is an extended method of Icinga Web 2 `ConfigForm`. You can' t add argument type to the child method without changing the parent method.

[the remark this answers is not in the document]
Comment (sukhwinder33445) [reply application/forms/RedisConfigForm.php:235]: Ahh, i see.

**Review by sukhwinder33445 — COMMENTED** [review]
LGTM!, Let @nilmerg have a look at it 😄 .

**Comment by Yonas-net** [comment]
> (I couldn't find anything regarding this on line)

Sure there is, the [PHP Manual](https://www.php.net/manual/de/language.types.declarations.php) states:
> When overriding a parent method, the child's method must match any return type declaration on the parent. If the parent 
> doesn't define a return type, then the child method may do so.

File: library/Icingadb/Common/IcingaRedis.php [hunk 3]
```diff
@@ -126,7 +133,7 @@ public static function getLastIcingaHeartbeat(Redis $redis = null)
         return null;
     }
 
-    public static function getPrimaryRedis(Config $moduleConfig = null, Config $redisConfig = null)
+    public static function getPrimaryRedis(Config $moduleConfig = null, Config $redisConfig = null): Redis
```
Comment (nilmerg) [inline library/Icingadb/Common/IcingaRedis.php:136]: `getSecondaryRedis` has a docstring, why is here none?

File: library/Icingadb/Common/ObjectInspectionDetail.php [hunk 4]
```diff
@@ -42,6 +42,11 @@ public function __construct(Model $object, array $apiResult)
         $this->joins = $apiResult['joins'];
     }
 
+    /**
+     * Render the object source location
+     *
+     * @return array|void
```
Comment (nilmerg) [inline library/Icingadb/Common/ObjectInspectionDetail.php:48]: `void` is only valid if there's not a single return statement, not even `return;` or `return null;`. So this should be `?array` instead.

File: library/Icingadb/Common/ObjectInspectionDetail.php [hunk 5]
```diff
@@ -58,6 +63,11 @@ protected function createSourceLocation()
         ];
     }
 
+    /**
+     * Render object's last check result
+     *
+     * @return array|void
```
Comment (nilmerg) [inline library/Icingadb/Common/ObjectInspectionDetail.php:69]: Same as above: `?array`

File: library/Icingadb/Common/ObjectInspectionDetail.php [hunk 6]
```diff
@@ -188,6 +198,13 @@ protected function createAttributes()
         ];
     }
 
+    /**
+     * Format the given value as a json
+     *
+     * @param $json
```
Comment (nilmerg) [inline library/Icingadb/Common/ObjectInspectionDetail.php:204]: Please add a type here. It's probably `mixed`.

File: library/Icingadb/Compat/CompatObject.php [hunk 7]
```diff
@@ -181,6 +181,13 @@ protected function getDataView()
         throw new NotImplementedError('getDataView() is not supported');
     }
 
+    /**
+     * Get the bool type of the given value as an int
+     *
+     * @param $value
```
Comment (nilmerg) [inline library/Icingadb/Compat/CompatObject.php:187]: Please add a type.

File: library/Icingadb/Compat/CompatObject.php [hunk 8]
```diff
@@ -181,6 +181,13 @@ protected function getDataView()
...
+     *
+     * @return int|void
```
Comment (nilmerg) [inline library/Icingadb/Compat/CompatObject.php:189]: `?int`

File: library/Icingadb/Compat/UrlMigrator.php [hunk 9]
```diff
@@ -8,6 +8,7 @@
 use ipl\Stdlib\Filter;
 use ipl\Web\Filter\QueryString;
 use ipl\Web\Url;
+use JetBrains\PhpStorm\Pure;
```
Comment (nilmerg) [inline library/Icingadb/Compat/UrlMigrator.php:11]: Huh? Please remove.

File: library/Icingadb/Compat/UrlMigrator.php [hunk 10]
```diff
@@ -84,9 +93,18 @@ public static function transformFilter(Filter\Rule $filter, $queryTransformer =
         }
 
         $rewritten = $transformer->rewrite($filter, $columns);
-        return $rewritten === false ? false : ($rewritten instanceof Filter\Rule ? $rewritten : $filter);
+        return $rewritten instanceof Filter\Rule ? $rewritten : $filter;
```
Comment (nilmerg) [inline library/Icingadb/Compat/UrlMigrator.php:96]: This changes the implementation. `rewrite` may still return `false`.

File: library/Icingadb/Compat/UrlMigrator.php [hunk 11]
```diff
@@ -84,9 +93,18 @@ public static function transformFilter(Filter\Rule $filter, $queryTransformer =
...
     }
 
+    /**
+     * Rewrite the given filter and legacy columns
+     *
+     * @param Filter\Rule $filter
+     * @param array $legacyColumns
+     * @param Filter\Chain|null $parent
+     *
+     * @return ?false|Filter\Condition|mixed
```
Comment (nilmerg) [inline library/Icingadb/Compat/UrlMigrator.php:106]: `...|mixed` is equivalent to `mixed`. In this case: `?mixed`

File: library/Icingadb/Hook/ActionsHook/ObjectActionsHook.php [hunk 12]
```diff
@@ -36,7 +37,7 @@ abstract class ObjectActionsHook
      *
      * @throws InvalidArgumentException If the given model is not supported
      */
-    final public static function loadActions(Model $object): HtmlElement
+    final public static function loadActions(Model $object): BaseHtmlElement
```
Comment (nilmerg) [inline library/Icingadb/Hook/ActionsHook/ObjectActionsHook.php:40]: It's a final method, here `HtmlElement` is fine I'd say.

File: library/Icingadb/Util/PerfData.php [hunk 13]
```diff
@@ -381,15 +381,15 @@ public function getWarningThreshold()
...
+    public function getCriticalThreshold(): ThresholdRange
     {
         return $this->criticalThreshold;
     }
 
     /**
      * Return the minimum value or null if it is not available
      *
-     * @return  null|string
+     * @return  null|float
```
Comment (nilmerg) [inline library/Icingadb/Util/PerfData.php:392]: `?float`?

File: library/Icingadb/Widget/Detail/ObjectsDetail.php [hunk 14]
```diff
@@ -34,14 +36,14 @@ class ObjectsDetail extends BaseHtmlElement
 
     protected $tag = 'div';
 
-    public function __construct($type, $summary, $query)
+    public function __construct(string $type, $summary, Query $query)
```
Comment (nilmerg) [inline library/Icingadb/Widget/Detail/ObjectsDetail.php:39]: Isn't `$summary` also a defined type?

File: library/Icingadb/Widget/Detail/ServiceStatistics.php [hunk 15]
```diff
@@ -7,7 +7,9 @@
 use Icinga\Chart\Donut;
 use Icinga\Data\Filter\Filter;
 use Icinga\Module\Icingadb\Common\Links;
+use Icinga\Module\Icingadb\Model\ServicestateSummary;
```
Comment (nilmerg) [inline library/Icingadb/Widget/Detail/ServiceStatistics.php:10]: Unused, please remove.

File: library/Icingadb/Widget/ItemList/BaseHostListItem.php [hunk 16]
```diff
@@ -23,6 +24,11 @@ abstract class BaseHostListItem extends StateListItem
 {
     use NoSubjectLink;
 
+    /**
+     * Create new subject link
+     *
+     * @return BaseHtmlElement|Link
```
Comment (nilmerg) [inline library/Icingadb/Widget/ItemList/BaseHostListItem.php:30]: `Link` is a `BaseHtmlElement`, so the latter suffices I think.

**Review by nilmerg — CHANGES_REQUESTED** [review]
Took a look at everything finally. :cold_sweat: Comments below :)

File: library/Icingadb/Compat/UrlMigrator.php (hunk 10 above)
Comment (Yonas-net) [reply library/Icingadb/Compat/UrlMigrator.php:96]: It doesn't matter what `rewrite()` is returning. If it isn't an instance of the `Filter\Rule`, the original filter will always be returned.

File: library/Icingadb/Compat/CompatObject.php [hunk 17]
```diff
@@ -181,6 +181,13 @@ protected function getDataView()
         throw new NotImplementedError('getDataView() is not supported');
     }
 
+    /**
+     * Get the bool type of the given value as an int
+     *
+     * @param bool $value
```
Comment (nilmerg) [inline library/Icingadb/Compat/CompatObject.php:187]: It's `bool|string`, isn't it?

File: library/Icingadb/Widget/Detail/ObjectsDetail.php [hunk 18]
```diff
@@ -34,14 +38,21 @@ class ObjectsDetail extends BaseHtmlElement
 
     protected $tag = 'div';
 
-    public function __construct($type, $summary, $query)
+    /**
+     * Construct an
```
Comment (nilmerg) [inline library/Icingadb/Widget/Detail/ObjectsDetail.php:42]: an what?

File: library/Icingadb/Widget/ItemList/BaseHostListItem.php [hunk 19]
```diff
@@ -23,6 +24,11 @@ abstract class BaseHostListItem extends StateListItem
 {
     use NoSubjectLink;
 
+    /**
+     * Create new subject link
+     *
+     * @return BaseHtmlElement
```
Comment (nilmerg) [inline library/Icingadb/Widget/ItemList/BaseHostListItem.php:30]: Cant' this be a return type declaration as well?

**Review by nilmerg — CHANGES_REQUESTED** [review]

File: library/Icingadb/Widget/ItemList/BaseHostListItem.php (hunk 19 above)
Comment (Yonas-net) [reply library/Icingadb/Widget/ItemList/BaseHostListItem.php:30]: Nope

File: library/Icingadb/Widget/Detail/ObjectsDetail.php (hunk 18 above)
Comment (Yonas-net) [reply library/Icingadb/Widget/Detail/ObjectsDetail.php:42]: 🙈

File: library/Icingadb/Compat/CompatObject.php (hunk 17 above)
Comment (Yonas-net) [reply library/Icingadb/Compat/CompatObject.php:187]: my bad

**Review by nilmerg — DISMISSED** [review]
:tada:

**Review by nilmerg — APPROVED** [review]

**Pull request merged**